File: reference/diff_crawler/classifier.py
"""File classification: decide how a given file should be diffed.

Categories
----------
code    -> textual unified diff (.py, .ipynb, .yaml, ...)
data    -> DuckDB row+schema diff (.csv, .parquet, .jsonl, ...)
binary  -> hash-only compare (.pkl, .pt, .h5, model weights, images, ...)
ignore  -> excluded entirely (venv, __pycache__, ...)
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

# --- Exclusion rules -------------------------------------------------------

# Directory names that are excluded wherever they appear in the path.
DEFAULT_EXCLUDED_DIRS: frozenset[str] = frozenset({
    "venv", ".venv", "env", ".env", "virtualenv",
    "__pycache__",
    ".git", ".hg", ".svn",
    "node_modules",
    ".ipynb_checkpoints",
    ".pytest_cache", ".mypy_cache", ".ruff_cache", ".tox",
    "dist", "build",
    "mlruns", "wandb", "lightning_logs",
    ".idea", ".vscode",
})

# Exact filenames to skip.
DEFAULT_EXCLUDED_FILES: frozenset[str] = frozenset({
    ".DS_Store", "Thumbs.db", ".gitignore.swp",
})

# File suffixes to skip outright.
DEFAULT_EXCLUDED_SUFFIXES: frozenset[str] = frozenset({
    ".pyc", ".pyo", ".pyd",
    ".class",
    ".log", ".tmp", ".swp", ".swo",
})

# --- Type rules ------------------------------------------------------------

CODE_SUFFIXES: frozenset[str] = frozenset({
    ".py", ".ipynb",
    ".js", ".ts", ".tsx", ".jsx",
    ".java", ".kt", ".scala",
    ".c", ".h", ".cpp", ".hpp", ".cc",
    ".go", ".rs", ".rb", ".php",
    ".sh", ".bash", ".zsh",
    ".r", ".jl",
    ".sql",
    ".yaml", ".yml", ".toml", ".ini", ".cfg",
    ".md", ".rst", ".txt",
    ".dockerfile", ".tf",
    ".html", ".css", ".scss",
})

DATA_SUFFIXES: frozenset[str] = frozenset({
    ".csv", ".tsv",
    ".parquet",
    ".jsonl", ".ndjson",
    ".feather", ".arrow",
    ".xlsx", ".xls",
})

# JSON is ambiguous: small -> treat as code/config, large -> treat as data.
JSON_DATA_THRESHOLD_BYTES = 256 * 1024  # 256 KB

# Binary / model weights — hash-compare only.
BINARY_SUFFIXES: frozenset[str] = frozenset({
    ".pkl", ".pickle", ".joblib",
    ".pt", ".pth", ".ckpt", ".safetensors",
    ".h5", ".hdf5",
    ".onnx", ".pb", ".tflite",
    ".npy", ".npz",
    ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".tiff", ".webp", ".svg",
    ".mp3", ".wav", ".flac", ".mp4", ".mov", ".avi",
    ".zip", ".tar", ".gz", ".bz2", ".7z",
    ".bin", ".dat",
})


@dataclass(frozen=True)
class Classification:
    kind: str          # "code" | "data" | "binary" | "ignore"
    reason: str        # short tag for the report


def is_excluded_path(rel_path: Path,
                     excluded_dirs: frozenset[str] = DEFAULT_EXCLUDED_DIRS,
                     excluded_files: frozenset[str] = DEFAULT_EXCLUDED_FILES,
                     excluded_suffixes: frozenset[str] = DEFAULT_EXCLUDED_SUFFIXES) -> bool:
    """Return True if rel_path should be skipped entirely."""
    parts = rel_path.parts
    if any(p in excluded_dirs for p in parts):
        return True
    if rel_path.name in excluded_files:
        return True
    if rel_path.suffix.lower() in excluded_suffixes:
        return True
    # Egg-info dirs use a suffix on a dir name
    if any(p.endswith(".egg-info") for p in parts):
        return True
    return False


def classify(abs_path: Path) -> Classification:
    """Classify a file by extension (and size for ambiguous types)."""
    suffix = abs_path.suffix.lower()

    if suffix == ".json":
        try:
            size = abs_path.stat().st_size
        except OSError:
            size = 0
        if size >= JSON_DATA_THRESHOLD_BYTES:
            return Classification("data", "json-large")
        return Classification("code", "json-config")

    if suffix in DATA_SUFFIXES:
        return Classification("data", f"data:{suffix.lstrip('.')}")

    if suffix in CODE_SUFFIXES:
        return Classification("code", f"code:{suffix.lstrip('.')}")

    if suffix in BINARY_SUFFIXES:
        return Classification("binary", f"binary:{suffix.lstrip('.')}")

    # Files with no extension or unknown extension: peek a few bytes.
    try:
        with open(abs_path, "rb") as f:
            chunk = f.read(2048)
    except OSError:
        return Classification("binary", "unreadable")

    if b"\x00" in chunk:
        return Classification("binary", "binary-sniffed")
    # Heuristic: mostly printable ASCII/UTF-8 -> treat as code
    try:
        chunk.decode("utf-8")
        return Classification("code", "text-sniffed")
    except UnicodeDecodeError:
        return Classification("binary", "non-utf8")

File: src/classifier.rs
use std::path::Path;

use crate::config::{EXCLUDED_DIRS, EXCLUDED_FILES, EXCLUDED_SUFFIXES};

/// How a file should be diffed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FileKind {
    Code,
    Data,
    Binary,
    Ignore,
}

#[derive(Debug, Clone)]
pub struct Classification {
    pub kind: FileKind,
    /// Short tag for diagnostics/logging.
    pub reason: &'static str,
}

// --- Extension sets (match classifier.py exactly) -------------------------

const CODE_SUFFIXES: &[&str] = &[
    ".py",
    ".ipynb",
    ".js",
    ".ts",
    ".tsx",
    ".jsx",
    ".java",
    ".kt",
    ".scala",
    ".c",
    ".h",
    ".cpp",
    ".hpp",
    ".cc",
    ".go",
    ".rs",
    ".rb",
    ".php",
    ".sh",
    ".bash",
    ".zsh",
    ".r",
    ".jl",
    ".sql",
    ".yaml",
    ".yml",
    ".toml",
    ".ini",
    ".cfg",
    ".md",
    ".rst",
    ".txt",
    ".dockerfile",
    ".tf",
    ".html",
    ".css",
    ".scss",
];

const DATA_SUFFIXES: &[&str] = &[
    ".csv", ".tsv", ".parquet", ".jsonl", ".ndjson", ".feather", ".arrow", ".xlsx", ".xls",
];

const BINARY_SUFFIXES: &[&str] = &[
    ".pkl",
    ".pickle",
    ".joblib",
    ".pt",
    ".pth",
    ".ckpt",
    ".safetensors",
    ".h5",
    ".hdf5",
    ".onnx",
    ".pb",
    ".tflite",
    ".npy",
    ".npz",
    ".png",
    ".jpg",
    ".jpeg",
    ".gif",
    ".bmp",
    ".tiff",
    ".webp",
    ".svg",
    ".mp3",
    ".wav",
    ".flac",
    ".mp4",
    ".mov",
    ".avi",
    ".zip",
    ".tar",
    ".gz",
    ".bz2",
    ".7z",
    ".bin",
    ".dat",
];

/// JSON files below this size are treated as code/config; at or above, as data.
const JSON_DATA_THRESHOLD_BYTES: u64 = 256 * 1024;

// --------------------------------------------------------------------------

/// Returns true if `rel` (a path relative to a tree root) should be skipped
/// entirely — not walked and not diffed.
///
/// Mirrors `is_excluded_path` in classifier.py. Called at walk time; the walker
/// also prevents descending into excluded directories via `filter_entry`, so in
/// practice this function acts as the per-file gate for suffix and filename checks.
pub fn is_excluded_path(rel: &Path) -> bool {
    for component in rel.components() {
        let s = component.as_os_str().to_string_lossy();
        if EXCLUDED_DIRS.contains(&s.as_ref()) {
            return true;
        }
        if s.ends_with(".egg-info") {
            return true;
        }
    }
    if let Some(name) = rel.file_name() {
        let name_str = name.to_string_lossy();
        if EXCLUDED_FILES.contains(&name_str.as_ref()) {
            return true;
        }
    }
    if let Some(ext) = rel.extension() {
        let suffix = format!(".{}", ext.to_string_lossy().to_lowercase());
        if EXCLUDED_SUFFIXES.contains(&suffix.as_str()) {
            return true;
        }
    }
    false
}

/// Classify a file by extension (and size for ambiguous types like `.json`).
///
/// Mirrors `classify()` in classifier.py. The sniffing fallback (null-byte /
/// UTF-8 probe) is used for files with no known extension.
pub fn classify(abs_path: &Path) -> Classification {
    let suffix = abs_path
        .extension()
        .map(|e| format!(".{}", e.to_string_lossy().to_lowercase()))
        .unwrap_or_default();

    if suffix == ".json" {
        let size = abs_path.metadata().map(|m| m.len()).unwrap_or(0);
        return if size >= JSON_DATA_THRESHOLD_BYTES {
            Classification {
                kind: FileKind::Data,
                reason: "json-large",
            }
        } else {
            Classification {
                kind: FileKind::Code,
                reason: "json-config",
            }
        };
    }

    if DATA_SUFFIXES.contains(&suffix.as_str()) {
        return Classification {
            kind: FileKind::Data,
            reason: "data",
        };
    }
    if CODE_SUFFIXES.contains(&suffix.as_str()) {
        return Classification {
            kind: FileKind::Code,
            reason: "code",
        };
    }
    if BINARY_SUFFIXES.contains(&suffix.as_str()) {
        return Classification {
            kind: FileKind::Binary,
            reason: "binary",
        };
    }

    // Unknown extension: sniff the first 2 KB.
    match sniff(abs_path) {
        Sniff::Binary => Classification {
            kind: FileKind::Binary,
            reason: "binary-sniffed",
        },
        Sniff::Code => Classification {
            kind: FileKind::Code,
            reason: "text-sniffed",
        },
        Sniff::Unreadable => Classification {
            kind: FileKind::Binary,
            reason: "unreadable",
        },
    }
}

enum Sniff {
    Binary,
    Code,
    Unreadable,
}

fn sniff(path: &Path) -> Sniff {
    use std::io::Read;
    let mut f = match std::fs::File::open(path) {
        Ok(f) => f,
        Err(_) => return Sniff::Unreadable,
    };
    let mut buf = [0u8; 2048];
    let n = match f.read(&mut buf) {
        Ok(n) => n,
        Err(_) => return Sniff::Unreadable,
    };
    let chunk = &buf[..n];
    if chunk.contains(&0u8) {
        return Sniff::Binary;
    }
    match std::str::from_utf8(chunk) {
        Ok(_) => Sniff::Code,
        Err(_) => Sniff::Binary,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[test]
    fn py_is_code() {
        let c = classify(Path::new("foo.py"));
        assert_eq!(c.kind, FileKind::Code);
    }

    #[test]
    fn parquet_is_data() {
        let c = classify(Path::new("data.parquet"));
        assert_eq!(c.kind, FileKind::Data);
    }

    #[test]
    fn pkl_is_binary() {
        let c = classify(Path::new("model.pkl"));
        assert_eq!(c.kind, FileKind::Binary);
    }

    #[test]
    fn exclude_venv_component() {
        assert!(is_excluded_path(Path::new("venv/lib/site.py")));
    }

    #[test]
    fn exclude_egg_info() {
        assert!(is_excluded_path(Path::new("mypackage.egg-info/PKG-INFO")));
    }

    #[test]
    fn exclude_pyc_suffix() {
        assert!(is_excluded_path(Path::new("foo.pyc")));
    }

    #[test]
    fn keep_normal_file() {
        assert!(!is_excluded_path(Path::new("src/main.py")));
    }

    #[test]
    fn keep_dotfile_without_excluded_suffix() {
        // .gitignore has no extension via Path::extension(), so not excluded by suffix rule.
        assert!(!is_excluded_path(Path::new(".gitignore")));
    }
}
